File: paciente_virtual/exames.py
"""Motores de exames: detectam solicitações do profissional e devolvem resultados do caso.

Os dois motores (exame físico e exames complementares) usam o mesmo critério:
o resultado só é entregue quando a frase contém uma expressão de solicitação
ativa ("vou aferir sua pressão", "solicito um ECG", "qual a saturação?").
Menção sem solicitação ("o senhor tem pressão alta?", "já fez um eletro?")
segue para o paciente como pergunta de anamnese.

O filtro é intencionalmente simples (palavras-chave, não interpretação):
frases como "costuma verificar sua pressão em casa?" ainda disparam a medição.
Casos ambíguos assim são raros e o custo é baixo; a alternativa (interpretar a
intenção com o LLM) tornaria o resultado não determinístico para avaliação.

Cada motor entrega TODOS os exames solicitados na frase e retorna a lista de
exames entregues (vazia se nenhum), para que a consulta informe o modelo de
linguagem do que aconteceu.
"""

import re

from .registro import TITULO_EXAME_FISICO, TITULO_EXAME_SOLICITADO, registrar
from .texto import contem_algum_termo, normalizar
from .voz.falar import falar

# Radicais de verbos de solicitação: casam qualquer flexão com \b<radical>\w*
# ("afer" cobre aferir/afira/aferindo; "solicit" cobre solicito/solicitar/solicite...).
_MARCADORES_RADICAL = [
    "afer",
    "verific",
    "chec",
    "avali",
    "examin",
    "auscult",
    "escut",
    "confer",
    "solicit",
    "realiz",
]

# Palavras e expressões exatas (com limite de palavra, acentos ignorados).
# "faz"/"faça" estão na lista, mas "fez" não: "já fez exame de pressão?" é
# pergunta de anamnese, não solicitação.
_MARCADORES_EXATOS = [
    "medir",
    "meça",
    "meço",
    "mede",
    "medindo",
    "tirar",
    "tire",
    "olhar",
    "olhe",
    "ver",
    "pedir",
    "peço",
    "faz",
    "faça",
    "fazer",
    "quero",
    "gostaria",
    "preciso",
    "qual",
    "quais",
    "quanto",
    "quanta",
    "como está",
    "como estão",
    "exame físico",
]

_PADRAO_SOLICITACAO = re.compile(
    "|".join(
        [rf"\b{radical}\w*" for radical in _MARCADORES_RADICAL]
        + [rf"\b{re.escape(normalizar(palavra))}\b" for palavra in _MARCADORES_EXATOS]
    )
)


def _ha_solicitacao(texto):
    return _PADRAO_SOLICITACAO.search(normalizar(texto)) is not None


def _termos_do_exame(chave, dados):
    termos = [chave, dados.get("nome", "")]
    termos.extend(dados.get("sinonimos", []))
    return [termo for termo in termos if termo]


def _entregar_resultado(titulo, dados, arquivo_historico, voz):
    print(f"\n{titulo}\n\n{dados['nome']}\n")
    print("Resultado:\n")
    print(dados["resultado"])
    print()

    falar(dados["resultado"], voz)

    registrar(arquivo_historico, f"\n{titulo}: {dados['nome']}\n")
    registrar(arquivo_historico, f"RESULTADO: {dados['resultado']}\n")


def _verificar(texto, exames, titulo, arquivo_historico, voz):
    if not exames or not _ha_solicitacao(texto):
        return []

    entregues = []
    for chave, dados in exames.items():
        if contem_algum_termo(texto, _termos_do_exame(chave, dados)):
            _entregar_resultado(titulo, dados, arquivo_historico, voz)
            entregues.append(dados)

    return entregues


def verificar_exames(texto, caso, arquivo_historico, voz):
    """Entrega os exames complementares solicitados na frase.

    Retorna a lista de exames entregues (vazia se nenhum).
    """
    return _verificar(
        texto,
        caso.get("exames_disponiveis"),
        TITULO_EXAME_SOLICITADO,
        arquivo_historico,
        voz,
    )


def verificar_exame_fisico(texto, caso, arquivo_historico, voz):
    """Entrega os itens de exame físico solicitados na frase.

    Retorna a lista de exames entregues (vazia se nenhum).
    """
    return _verificar(
        texto,
        caso.get("exame_fisico"),
        TITULO_EXAME_FISICO,
        arquivo_historico,
        voz,
    )

File: deploy/hostinger/motor/exames.js
// Detecção de solicitações de exame.
//
// Um achado só é entregue quando o profissional REALIZA ou SOLICITA o exame agora —
// perguntar da história ("qual a sua pressão normalmente em casa?") é anamnese, não
// aferição, e não pode entregar o valor medido nem contar ponto na rubrica.
//
// Por isso há dois níveis de marcador:
//   AÇÃO   — verbo de procedimento ("vou aferir", "solicito", "aplicar"). Dispara sempre.
//   CONSULTA — pergunta pelo valor ("qual", "como está"). Dispara só se a frase NÃO
//              estiver falando do passado/do hábito ("normalmente", "em casa", "já").

import { contemAlgumTermo, normalizar } from "./texto.js";

export const TITULO_EXAME_FISICO = "EXAME FÍSICO";
export const TITULO_EXAME_SOLICITADO = "EXAME SOLICITADO";

// Sinais vitais / medidas rápidas — separados dos demais achados do exame físico.
// Vive aqui (e não no servidor) porque tanto a UI quanto o pedido em bloco
// ("vou verificar os sinais vitais") precisam da mesma lista.
export const CHAVES_VITAIS = new Set([
  "pressao_arterial",
  "frequencia_cardiaca",
  "frequencia_respiratoria",
  "temperatura",
  "saturacao",
  "glicemia_capilar",
  "glicemia",
  "escala_dor",
  "dor",
  "peso",
]);

// Radicais de verbos de procedimento: casam qualquer flexão com \b<radical>\w*.
const ACAO_RADICAL = [
  "afer",
  "verific",
  "chec",
  "avali",
  "examin",
  "auscult",
  "escut",
  "confer",
  "solicit",
  "realiz",
  "aplic",
  "observ",
  "palp",
  "percut",
  "inspecion",
];

// Verbos/expressões de procedimento sem radical comum.
const ACAO_EXATA = [
  "medir",
  "meça",
  "meço",
  "mede",
  "medindo",
  "tirar",
  "tire",
  "pedir",
  "peço",
  "faça",
  "fazer",
  "faz",
  "quero",
  "gostaria",
  "preciso",
  "vamos ver",
  "deixa eu ver",
  "deixe-me ver",
  "exame físico",
];

// Perguntas pelo valor: só valem como solicitação no aqui-e-agora.
const CONSULTA_EXATA = ["qual", "quais", "quanto", "quanta", "como está", "como estão", "ver", "olhar", "olhe"];

// Marcadores de história/hábito. Se aparecerem, uma pergunta do tipo CONSULTA é
// anamnese ("qual sua pressão normalmente?") e não entrega o valor aferido.
const HISTORIA = [
  "normalmente",
  "geralmente",
  "habitualmente",
  "em geral",
  "de costume",
  "costuma",
  "costumava",
  "em casa",
  "já",
  "ja",
  "alguma vez",
  "antes",
  "no passado",
  "outro dia",
  "última vez",
  "ultima vez",
  "toda vez",
  "sempre",
  "anteriormente",
  "recentemente",
];

// Termos que pedem o conjunto de sinais vitais de uma vez.
const TERMOS_VITAIS = ["sinais vitais", "sinais", "vitais", "dados vitais"];

function escaparRegex(texto) {
  return texto.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

function padraoDe(radicais, exatos) {
  return new RegExp(
    [
      ...radicais.map((radical) => `\\b${radical}\\w*`),
      ...exatos.map((palavra) => `\\b${escaparRegex(normalizar(palavra))}\\b`),
    ].join("|")
  );
}

const PADRAO_ACAO = padraoDe(ACAO_RADICAL, ACAO_EXATA);
const PADRAO_CONSULTA = padraoDe([], CONSULTA_EXATA);
const PADRAO_HISTORIA = padraoDe([], HISTORIA);

// True quando a frase pede um procedimento AGORA (e não conta a história dele).
export function haSolicitacao(texto) {
  const t = normalizar(texto);
  if (PADRAO_ACAO.test(t)) return true;
  return PADRAO_CONSULTA.test(t) && !PADRAO_HISTORIA.test(t);
}

function termosDoExame(chave, dados) {
  return [chave, dados.nome || "", ...(dados.sinonimos || [])].filter(Boolean);
}

function detectar(texto, exames) {
  if (!exames || !haSolicitacao(texto)) return [];
  return Object.entries(exames)
    .filter(([chave, dados]) => contemAlgumTermo(texto, termosDoExame(chave, dados)))
    .map(([, dados]) => dados);
}

// "Vou verificar os sinais vitais" pede o bloco inteiro — sem isto o pedido mais
// natural da beira do leito não entregava nada.
function detectarVitais(texto, exameFisico) {
  if (!exameFisico || !haSolicitacao(texto)) return [];
  if (!contemAlgumTermo(texto, TERMOS_VITAIS)) return [];
  return Object.entries(exameFisico)
    .filter(([chave, dados]) => CHAVES_VITAIS.has(chave) && dados && typeof dados === "object")
    .map(([, dados]) => dados);
}

// Retorna pares [titulo, dados] — complementares primeiro, exame físico depois.
export function detectarExames(texto, caso) {
  const fisico = [...detectarVitais(texto, caso.exame_fisico), ...detectar(texto, caso.exame_fisico)];
  // Um pedido pode casar duas vezes (bloco de vitais + nome do sinal): mantém a
  // primeira ocorrência para não repetir o mesmo achado na tela e no transcript.
  const vistos = new Set();
  const fisicoUnico = fisico.filter((dados) => {
    const chave = dados.nome || JSON.stringify(dados);
    if (vistos.has(chave)) return false;
    vistos.add(chave);
    return true;
  });

  return [
    ...detectar(texto, caso.exames_disponiveis).map((dados) => [TITULO_EXAME_SOLICITADO, dados]),
    ...fisicoUnico.map((dados) => [TITULO_EXAME_FISICO, dados]),
  ];
}
